File: projects/ios-device-agent/IOSDeviceAgent/DoguCommon/SizePrefixedRecvQueue.swift
//
//  SizePrefixedRecvQueue.swift
//  DoguScreen
//
//  Created by jenkins on 2023/06/07.
//  Copyright © 2023 Dogu. All rights reserved.
//

import Foundation

class SizePrefixedRecvQueue {
  private var buffer: Data

  init() {
    buffer = Data()
  }

  func pushBuffer(buffer: Data) {
    var newBuffer = Data(count: self.buffer.count + buffer.count)
    newBuffer[0..<self.buffer.count] = self.buffer
    newBuffer[self.buffer.count..<newBuffer.count] = buffer
    self.buffer = newBuffer
  }

  func has() -> Bool {
    if buffer.count < 4 {
      return false
    }
    let packetSize = readUInt32(data: buffer, offset: 0)
    if buffer.count < packetSize + 4 {
      return false
    }
    return true
  }

  func pop() -> Data {
    assert(has(), "PacketQueue no package to pop")

    let packetSize = readUInt32(data: buffer, offset: 0)
    let ret = buffer.subdata(in: 4..<4 + Int(packetSize))
    buffer = buffer.subdata(in: 4 + Int(packetSize)..<buffer.count)
    return ret
  }

  private func readUInt32(data: Data, offset: Int) -> UInt32 {
    let subdata = data.subdata(in: offset..<offset + MemoryLayout<UInt32>.size)
    return subdata.withUnsafeBytes { $0.load(as: UInt32.self) }
  }
}

File: packages/typescript/common/src/queue/size-prefixed-recv-queue.ts
import assert from 'assert';
import { Uint8ArrayUtil } from '..';

export class SizePrefixedRecvQueue {
  private buffer: Uint8Array;

  constructor() {
    this.buffer = new Uint8Array(0);
  }

  public pushBuffer(buffer: Uint8Array): void {
    const newBuffer = new Uint8Array(this.buffer.length + buffer.length);
    newBuffer.set(this.buffer);
    newBuffer.set(buffer, this.buffer.length);
    this.buffer = newBuffer;
  }

  public popLoop(callback: (array: Uint8Array) => void): void {
    for (let i = 0; i < 100000; i++) {
      if (!this.has()) {
        return;
      }
      const array = this.pop();
      callback(array);
    }
  }

  private has(): boolean {
    if (this.buffer.length < 4) return false;
    const packetSize = Uint8ArrayUtil.readUint32(this.buffer, 0);
    if (this.buffer.length < packetSize + 4) return false;
    return true;
  }

  private pop(): Uint8Array {
    assert(this.has(), 'PacketQueue no package to pop');

    const packetSize = Uint8ArrayUtil.readUint32(this.buffer, 0);
    const ret = this.buffer.subarray(4, 4 + packetSize);
    this.buffer = this.buffer.subarray(4 + packetSize, this.buffer.length);
    return ret;
  }

  public clear(): void {
    this.buffer = new Uint8Array(0);
  }
}
